source

```diff
@@ -88993,45 +88999,55 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
      </li>
 
      <li>
-      <p>If there are <span data-x="top-level browsing context">top-level browsing contexts</span>
-      <var>B</var> that the user agent believes would not benefit from having their rendering
-      updated at this time, then remove from <var>docs</var> all <code>Document</code> objects whose
-      <span data-x="concept-document-bc">browsing context</span>'s <span>top-level browsing
-      context</span> is in <var>B</var>.</p>
 
-      <div class="note">
-       <p>Whether a <span>top-level browsing context</span> would benefit from having its rendering
-       updated depends on various factors, such as the update frequency. For example, if the browser
-       is attempting to achieve a 60Hz refresh rate, then these steps are only necessary every 60th
-       of a second (about 16.7ms). If the browser finds that a <span>top-level browsing
-       context</span> is not able to sustain this rate, it might drop to a more sustainable 30Hz for
-       that set of <code>Document</code>s, rather than occasionally dropping frames. (This
-       specification does not mandate any particular model for when to update the rendering.)
-       Similarly, if a <span>top-level browsing context</span> is in the background, the user agent
-       might decide to drop that page to a much slower 4Hz, or even less.</p>
-
-       <p>Another example of why a browser might skip updating the rendering is to ensure certain
-       <span data-x="concept-task">tasks</span> are executed immediately after each other, with only
-       <span data-x="perform a microtask checkpoint">microtask checkpoints</span> interleaved (and
-       without, e.g., <span data-x="run the animation frame callbacks">animation frame
-       callbacks</span> interleaved). For example, a user agent might wish to coalesce timer
-       callbacks together, with no intermediate rendering updates.</p>
-      </div>
+      <p>If there are a <span data-x="browsing context">browsing contexts</span> <var>B</var> that
+      do not have a rendering opportunity, then remove from <var>docs</var> all
+      <code>Document</code> objects whose <span data-x="concept-document-bc">browsing context</span>
+      is in <var>B</var>.</p>
+
+      <p>A <span>browsing context</span> has a rendering opportunity if the user agent is currently
+      able to present the contents of the <span>browsing context</span> to the user, accounting for
+      hardware refresh rate constraints and user agent throttling for performance reasons, but
+      considering content presented even if it's outside the viewport.</p>
```

Comment:
This is messy, but I haven't come up with anything any better. Any thoughts?

Reply:
I'm generally quite impressed by how this is all phrased. I don't find this messy.

However, I'm unsure of the practical implications of "considering content presented even if it's outside the viewport". I guess it's saying that iframes are supposed to "have rendering opportunities" even if they're below the fold?

It seems like by this phrasing it'd be OK to throttle offscreen iframes down to 0 FPS "for performance reasons", but would not be OK to consider them to "not have a rendering opportunity" just because they're offscreen. Which is a bit weird, but maybe that's just because limiting behaviors are often weird.

Maybe it'd help to explicitly connect this to the below step, which allows removing for a wider variety of reasons. E.g. explicitly stating that this step just establishes "rendering opportunities", and the below step is where the browser decides whether to take advantage of the opportunity or not. Or maybe explicitly mentioning the term "frame rate".

Reply:
I've tried to clarify this. I also realized I was incorrectly handling the case where we coalesce tasks. We shouldn't mark paint timing in that case, but are doing so in this spec.